apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.SignedPostPolicy.cs

```diff
@@ -0,0 +1,83 @@
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Cloud.Storage.V1
+{
+    public sealed partial class UrlSigner
+    {
+        /// <summary>
+        /// The result of signing a <see cref="PostPolicy"/>
+        /// containing the fields that should be including when
+        /// making the post request.
+        /// </summary>
+        public sealed class SignedPostPolicy
+        {
+            private const string SignatureElementName = "x-goog-signature";
+            private const string PolicyElementName = "policy";
+
+            internal SignedPostPolicy(PostPolicy source, string encodedPolicy, string signature, Uri url)
+            {
+                Source = GaxPreconditions.CheckNotNull(source, nameof(source));
+                PostUrl = GaxPreconditions.CheckNotNull(url, nameof(url));
+                EncodedPolicy = GaxPreconditions.CheckNotNull(encodedPolicy, nameof(encodedPolicy));
+                Signature = GaxPreconditions.CheckNotNull(signature, nameof(signature));
+            }
+
+            internal PostPolicy Source { get; }
+
+            internal string EncodedPolicy { get; }
+
+            internal string Signature { get; }
+
+            /// <summary>
+            /// The URL where the post request enforced by this signed policy should be made.
+            /// </summary>
+            public Uri PostUrl { get; }
+
+            /// <summary>
+            /// Collection of fields as defined by this policy.
+            /// Post requests enforced by this policy should set, for each
+            /// pair in this sequence, the given element to the exact given value.
+            /// </summary>
+            public IEnumerable<(string element, string value)> Fields =>
+                Source.Fields.Concat(new (string element, string value)[] 
+                {
+                    (SignatureElementName, Signature),
+                    (PolicyElementName, EncodedPolicy)
+                });
+
+            /// <summary>
+            /// Collection of start with conditions as defined by this policy.
+            /// Post requests enforced by this policy should set, for each
+            /// pair in this sequence, the given element to a value starting with the given prefix.
+            /// </summary>
+            public IEnumerable<(string element, string prefix)> StartWithConditions =>
```

Comment:
Could you provide more information on what this helper does?

Reply:
From what I understand one of the main use cases of post policies is as follows:

- I have a service that as part of its functionality allows consumers (other apps) to upload files to a bucket.
- My service emits post policies to those other apps to constraint the posts.
- The other apps use the policy to generate, show and validate a form to the final user who chooses the file they want to post.
- If only the fields are returned by the policy, the other apps can only partially validate that whatever the user is posting matches the policy.
- The other apps need also to know the conditions to validate that what the user is uploading is valid before it reaches the server.

Reply:
Gotcha, thanks for the explanation @amanda-tarafa,

I'd hold off on surfacing *Conditions as return values in the policy. I understand the use-case to add them but it wasn't an expectation for implementation and C# would be the only one to do it at the moment.

I'm going to side on hearing customer feedback before introducing more complexity to the surface.

Please revert this surface.

Reply:
I've remove them, but I honestly think we are only partially helping post policy consumers if we don't include the conditions.